### PR TITLE
Add stats update endpoint.

### PR DESCRIPTION
**Pull request by fabricematrat**

### CODE REVIEW

**Comment by fabricematrat** [comment]
QA:

Add a charm to the store.
call the /stats/update endpoint for that charm by looking at the docs API.md (will help to validate the docs too)
Check the download stats have increased by one by calling /pathtocharm/meta/stats?refresh=1

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/charmstore/1117/
Test PASSed.

File: internal/v4/api_test.go [hunk 1]
```diff
@@ -2246,7 +2246,7 @@ func (s *APISuite) TestWhoAmIFailWithNoMacaroon(c *gc.C) {
 func (s *APISuite) TestWhoAmIReturnsNameAndGroups(c *gc.C) {
 	s.discharge = dischargeForUser("who")
 	s.idM.groups = map[string][]string{
-		"who": []string{"foo", "bar"},
+		"who": {"foo", "bar"},
```
Comment (mhilton) [inline internal/v4/api_test.go]: does this compile on 1.2?

File: internal/v4/stats.go [hunk 2]
```diff
@@ -11,8 +12,8 @@ import (
 
 	"gopkg.in/errgo.v1"
 	"gopkg.in/juju/charmrepo.v1/csclient/params"
-
```
Comment (mhilton) [inline internal/v4/stats.go:14]: This blank should still be there.

File: internal/v4/stats.go [hunk 3]
```diff
@@ -111,6 +112,39 @@ func (h *ReqHandler) serveStatsCounter(_ http.Header, r *http.Request) (interfac
...
+	}
+
+	// IncrementDownloadCountsAtTime will check if the entity is found
+	// and if this is or not a promulgated entity
+	id := &router.ResolvedURL{
+		URL:                 *req.CharmReference,
+		PromulgatedRevision: -1,
+	}
+
+	logger.Infof("Increase donwload stats for id: %s at time: %s", id, req.Timestamp)
```
Comment (mhilton) [inline internal/v4/stats.go]: *download

File: internal/v4/stats.go [hunk 4]
```diff
@@ -111,6 +112,39 @@ func (h *ReqHandler) serveStatsCounter(_ http.Header, r *http.Request) (interfac
 	return items, nil
 }
 
+// PUT stats/update
+// https://github.com/juju/charmstore/blob/v4/docs/API.md#put-statsupdate
+func (h *ReqHandler) serveStatsUpdate(_ http.Header, r *http.Request) (interface{}, error) {
```
Comment (mhilton) [inline internal/v4/stats.go:117]: Surely this should be an admin only operation?

**Comment by mhilton** [comment]
generally good, but I think you need to restrict the endpoint with an ACL.

**Comment by jujugui** [comment]
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/charmstore/1118/
Test FAILed.

**Comment by jujugui** [comment]
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/charmstore/1119/
Test FAILed.

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/charmstore/1120/
Test PASSed.

File: internal/v4/api_test.go (hunk 1 above)
Comment (fabricematrat) [inline internal/v4/api_test.go]: Done.

File: internal/v4/stats.go (hunk 2 above)
Comment (fabricematrat) [inline internal/v4/stats.go:14]: Done.

File: internal/v4/stats.go (hunk 3 above)
Comment (fabricematrat) [inline internal/v4/stats.go]: Done.

File: internal/v4/stats.go (hunk 4 above)
Comment (fabricematrat) [inline internal/v4/stats.go:117]: Done.

**Comment by urosj** [comment]
:+1: QA as well.

**Comment by mhilton** [comment]
:+1: LGTM

**Comment by fabricematrat** [comment]
:shipit: thanks @urosj && @mhilton

**Comment by jujugui** [comment]
Status: merge request accepted. Url: http://ci.jujugui.org:8080/job/charmstore-merge

**Pull request merged**